CIP-transaction-inputs-as-unordered-set/README.md

```diff
@@ -0,0 +1,140 @@
+---
+CIP: xxx
+Title: Transaction Inputs as Unordered Set
```

Comment:
There is a bit of terminology inconsistency IMHO.

Unordered set is usually a set that not only does not preserve the order, but it also doesn't have any well defined order. In the matter of fact CDDL's [`set`](https://github.com/input-output-hk/cbor-sets-spec/blob/master/CBOR_SETS.md#introduction) is such a type: 

> order has no significance

Other examples are a [HashSet](https://hackage.haskell.org/package/unordered-containers-0.2.20/docs/Data-HashSet.html#t:HashSet) is an unordered set in Haskell or [`unordered_set`](https://en.cppreference.com/w/cpp/container/unordered_set) in C++.

What we have already implemented in ledger is ordered sets. Commonly an ordered set in math refers to a set with a specified ordering on elements. However, CDDL does not really have a notion of ordering outside of canonical CBOR, which isn't very useful for Cardano anyways. So, I believe it is safe to refer to an `oset` in CDDL as a Set that preserves the order of elements that was supplied in the serialized form of a transaction, which seems like exactly what you are looking for in this CIP

```suggestion
Title: Transaction Inputs as an Ordered Set
```

Reply:
thanks @lehins for giving voice to my own feelings - @solidsnakedev I was also wondering why the term "unordered" was being used to impose constraints on ordering.

We've decided at the last CIP meeting to assign this a number & traditionally this is when we also confirm the title (since it appears in the repository `README`) so please post ASAP with reasons for any objection you may have, and otherwise I'll change it as recommended above in both the PR & document title.

Reply:
Maybe seeing it from a programmers perspective rather than from math the term "Unique list" would be less ambiguous? i.e. a list with only unique elements.

Reply:
I wish we could have had engagement from @solidsnakedev about the title issue before promoting this to a candidate, but in the meantime I have to do this with a best-guess title that I think is accurate & least likely to be modified later.

Therefore am changing to the title that @lehins recommended, which would be especially important given the corresponding code change recommendation (https://github.com/cardano-foundation/CIPs/pull/758#discussion_r1690173711).  The life cycle of this CIP depends upon Ledger  support so I'd give first priority to whatever title the Ledger team would prefer to assign to the data structures _they_ ultimately maintain.

Obviously some rewrite of the CIP text would then also be required since the term "unordered" is consistently used to mean "not lexicographically reordered" and so we'll leave this thread unresolved until the author & readers completely agree on a way of addressing that ambiguity.

Reply:
Currently transaction outputs are an unordered set in the CDDL, and semantically this means that you can put them in whatever order you want while constructing the transaction and that order will be preserved in the script context. That is why the name unordered set was chosen for this, to be consistent with the naming of the type in the CDDL for tx outputs.

Reply:
@colll78 as I was updating the title I also considered suggesting:
> Preserving Order of Transaction Inputs

Would that be accurate & less ambiguous?  If not, could it be modified a bit to explain completely... _in terms of what the CIP provides_, rather than how it relates to internal data structures?  cc @solidsnakedev @lehins

Reply:
I totally support this title: "Preserving Order of Transaction Inputs"

It removes the concern and decouples the CIP title from terminology used in the implementation

Reply:
> That is why the name unordered set was chosen for this

@colll78 I understand the motivation and reasoning for choosing this name, but the name of unordered set means something totally different in programming. It means a Set with these properties:
* amortized constant time for lookups
* inserts and deletes also have amortized constant time
* order of inserts is not preserved!

In other words Hash Set is a true unordered set, but it is absolutely opposite from what is desired in this CIP.

Reply:
OK, I'm changing it to `Preserving Order of Transaction Inputs` since it either addresses or evades all the points made in the conversation above (so resolving this thread).